### PR TITLE
Initial commit of NumberBox sample.

### PR DESCRIPTION
**Pull request by SavoySchuler**
## Description

Added NumberBox sample. 

## Motivation and Context

New WinUI control releasing with WinUI 2.3.

## How Has This Been Tested?

Manual sample testing on Windows 10 Enterprise Insider Preview.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by SavoySchuler** [comment]
@stmoy I noticed VariableSizeWrapGrid had a dynamic code snippet that updated when a new configuration option was applied. However, this was written in the old in-XAML format rather than with the new rich sample code coloring feature. Is there a pre-defined path for dynamically updating these new code snippets?

**Comment by chingucoding** [comment]
> Is there a pre-defined path for dynamically updating these new code snippets?

I think the ControlExample.Substitutions feature is the way to achieve that. The substitutions both work with inline XAML and code samples loaded from files.

Substitutions in Button page (which use inline XAML):
https://github.com/microsoft/Xaml-Controls-Gallery/blob/53116b7e736a2a9d8a96e25a1e7fc5b7cc49ad4b/XamlControlsGallery/ControlPages/ButtonPage.xaml#L38-L41

Substitions in ListView page (with source file):
https://github.com/microsoft/Xaml-Controls-Gallery/blob/53116b7e736a2a9d8a96e25a1e7fc5b7cc49ad4b/XamlControlsGallery/ControlPages/ListViewPage.xaml#L260-L262

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿<controls:NumberBox x:Name="NumberBoxSpinButtonPlacementExample"
+    Header="Enter an integer:" 
+    Value="1" 
+    SpinButtonPlacementMode="Compact" 
+    StepFrequency="1" >
+</controls:NumberBox>
```
[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt]: Thanks, @chingucoding. I was following the TeachingTip sample, but it indeed looks like we have changed to the "muxc" namespace for new controls. I'll file a bug for myself on the TeachingTip sample so that I get that updated as well.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+﻿<controls:NumberBox x:Name="FormattedNumberBox" Header="Enter an dollar amount:" PlaceholderText="0.00" />
```
[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt]: Fixed! Thanks for pointing it out! 😊

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt]: Fixed!

File: XamlControlsGallery/DataModel/ControlInfoData.json [hunk 3]
```diff
@@ -1851,6 +1851,31 @@
...
+          "Content": "<p>Look at the <i>NumberBox.xaml</i> file in Visual Studio to see the full code for this page.</p>",
+          "IsNew": true,
+          "Docs": [
+            {
+              "Title": "NumberBox - API",
+              "Uri": "https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.numberbox"
+            },
+            {
+              "Title": "Guidelines",
+              "Uri": "https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/number-box"
```
[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/DataModel/ControlInfoData.json]: Nice catch! Fixed!

[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/DataModel/ControlInfoData.json]: And the URL will be going live this week. 😊

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+﻿<controls:NumberBox Header="Enter an expression:" Value="NaN" PlaceholderText="1 + 2^2" AcceptsExpression="True" />
```
[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt]: Corrected the namespace to "muxc".

@stmoy - followed this "new file for sample code" template to make sure I got your rich text features. Can I also get that with a sample inlined to the page? I follow your preferred path here.

[the remark this answers is not in the document]
Comment (stmoy) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt]: @SavoySchuler - @chingucoding is correct. Please move this single line to be inline in the control example. An example is here: [Button](https://github.com/microsoft/Xaml-Controls-Gallery/blob/53116b7e736a2a9d8a96e25a1e7fc5b7cc49ad4b/XamlControlsGallery/ControlPages/ButtonPage.xaml#L52)

[the remark this answers is not in the document]
Comment (SavoySchuler) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt]: @stmoy Would you prefer this for the other examples too or just the one-line example here? The others are also relatively simple.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿<muxc:NumberBox x:Name="NumberBoxSpinButtonPlacementExample"
+    Header="Enter an integer:" 
+    Value="1" 
+    SpinButtonPlacementMode="Compact" 
+    SmallChange="10"
+	LargeChange="100" >
+</muxc:NumberBox>
```
Comment (YuliKl) [inline XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt]: This doesn't have @chingucoding's original suggestion:
> Nit: " />" instead of "</..>

And now the LargeChange got too indented :)

File: XamlControlsGallery/DataModel/ControlInfoData.json [hunk 6]
```diff
@@ -1851,6 +1851,31 @@
             "TextBox"
           ]
         },
+        {
+          "UniqueId": "NumberBox",
+          "Title": "NumberBox",
+          "Subtitle": "A single-line or multi-line plain text field.",
```
Comment (YuliKl) [inline XamlControlsGallery/DataModel/ControlInfoData.json]: Sounds like this was copied from TextBox, doesn't seem entirely accurate for NumberBox.

File: XamlControlsGallery/DataModel/ControlInfoData.json [hunk 7]
```diff
@@ -1851,6 +1851,31 @@
...
+          "ImagePath": "ms-appx:///Assets/NumberBox.png",
+          "Description": "Use a NumberBox to let a user enter numeric and formulaic input in your app.",
```
Comment (YuliKl) [inline XamlControlsGallery/DataModel/ControlInfoData.json]: I may be overthinking, but "formulaic input" doesn't sound right to me.

File: XamlControlsGallery/ControlPages/NumberBox.xaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        <local:ControlExample HeaderText="A NumberBox with a spin button." 
+                              XamlSource="NumberBox/NumberBoxSample2_xaml.txt">
+            <local:ControlExample.Example>
+                <StackPanel>
+                    <muxc:NumberBox x:Name="NumberBoxSpinButtonPlacementExample"
+                                    Header="Enter an integer:" 
+                                    Value="10" 
+                                    SpinButtonPlacementMode="Compact" 
+                                    SmallChange="10"
+                                    LargeChange="100">
```
Comment (YuliKl) [inline XamlControlsGallery/ControlPages/NumberBox.xaml]: Same request as in the code sample: please close the NumberBox inline instead of having the hanging `</muxc:NumberBox>` on its own line.

File: XamlControlsGallery/ControlPages/NumberBox.xaml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                              XamlSource="NumberBox/NumberBoxSample1_xaml.txt">
+            <local:ControlExample.Example>
+                <muxc:NumberBox Header="Enter an expression:" Value="NaN" PlaceholderText="1 + 2^2" AcceptsExpression="True" />
+            </local:ControlExample.Example>
+        </local:ControlExample>
+
+        <local:ControlExample HeaderText="A NumberBox with a spin button." 
+                              XamlSource="NumberBox/NumberBoxSample2_xaml.txt">
+            <local:ControlExample.Example>
+                <StackPanel>
```
Comment (YuliKl) [inline XamlControlsGallery/ControlPages/NumberBox.xaml]: I don't think you need this stack panel.

File: XamlControlsGallery/ControlPages/NumberBox.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                switch (spinButtonPlacementModeName)
+                {
+                    case "Inline":
+                        NumberBoxSpinButtonPlacementExample.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Inline;
+                        break;
+                    case "Compact":
+                        NumberBoxSpinButtonPlacementExample.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Compact;
+                        break;
+                }
+
```
Comment (YuliKl) [inline XamlControlsGallery/ControlPages/NumberBox.xaml.cs]: nit: there's unnecessary whitespace here that I think makes the the code harder to read.

File: XamlControlsGallery/ControlPages/NumberBox.xaml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" 
+        xmlns:local="using:AppUIBasics"
+        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+        xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
+        mc:Ignorable="d">
+
+    <StackPanel HorizontalAlignment="Left">
+
```
Comment (stmoy) [inline XamlControlsGallery/ControlPages/NumberBox.xaml]: Nit: Please remove unnecessary space.

File: XamlControlsGallery/ControlPages/NumberBox.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
+// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
+// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
+//
+//*********************************************************C:\Users\saschule\Source\Repos\Xaml-Controls-Gallery\XamlControlsGallery\Assets\InkCanvas.png
+using Microsoft.UI.Xaml.Controls;
+using Windows.Globalization.NumberFormatting;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+// The Blank Page item template is documented at http://go.microsoft.com/fwlink/?LinkId=234238
```
Comment (stmoy) [inline XamlControlsGallery/ControlPages/NumberBox.xaml.cs]: Please remove this boilerplate.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+﻿<muxc:NumberBox Header="Enter an expression:" Value="NaN" PlaceholderText="1 + 2^2" AcceptsExpression="True" />
```
Comment (stmoy) [inline XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample1_xaml.txt]: Please move this code inline into the sample since this is a one-liner text file.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt (hunk 5 above)
Comment (stmoy) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample2_xaml.txt]: The trouble here is tabs vs. spaces. The whitespace before the LargeChange is a tab, but should be spaces. @SavoySchuler please ensure that we're using spaces.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt [hunk 14]
```diff
@@ -0,0 +1 @@
+﻿<muxc:NumberBox x:Name="FormattedNumberBox" Header="Enter an dollar amount:" PlaceholderText="0.00" />
```
Comment (stmoy) [inline XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt]: As above, please move this sample inline.

File: XamlControlsGallery/DataModel/ControlInfoData.json (hunk 7 above)
Comment (stmoy) [reply XamlControlsGallery/DataModel/ControlInfoData.json]: I mean... it does support formulaic input, right?

File: XamlControlsGallery/DataModel/ControlInfoData.json [hunk 15]
```diff
@@ -1851,6 +1851,31 @@
...
+              "Uri": "https://docs.microsoft.com/uwp/api/windows.ui.xaml.controls.numberbox"
+            },
+            {
+              "Title": "Guidelines",
+              "Uri": "https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/number-box"
+            }
+          ],
+          "RelatedControls": [
+            "TextBox",
+            "TextBlock",
```
Comment (stmoy) [inline XamlControlsGallery/DataModel/ControlInfoData.json]: What is the relationship between NumberBox and TextBlock?

File: XamlControlsGallery/XamlControlsGallery.csproj [hunk 16]
```diff
@@ -1105,19 +1117,19 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="ColorCode.UWP">
-      <Version>2.0.5</Version>
```
Comment (stmoy) [inline XamlControlsGallery/XamlControlsGallery.csproj]: This change shouldn't rev the versions for unrelated packages. Please revert these changes.

**Comment by stmoy** [comment]
@SavoySchuler - I left some feedback above, please let me know if anything is unclear.

Two more quick things:

1. It looks like this change has a merge conflict - please resolve the conflict and recommit
2. Please include some screenshots in the original description above. Screenshots are very helpful because this is a new sample.

File: XamlControlsGallery/DataModel/ControlInfoData.json (hunk 7 above)
Comment (YuliKl) [reply XamlControlsGallery/DataModel/ControlInfoData.json]: definition of formulaic: "constituting or containing a verbal formula or set form of words; produced in accordance with a slavishly followed rule or style; predictable."

A better description might be: Use NumberBox to allow users to enter numeric input and evaluate algebraic  equations in your app.

File: XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt (hunk 14 above)
Comment (stmoy) [reply XamlControlsGallery/ControlPagesSampleCode/NumberBox/NumberBoxSample3_xaml.txt]: Since this sample also has a CS txt file, I can see an argument to keep the xaml in a txt file as well.
